x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/Match.java

```diff
@@ -52,10 +80,14 @@ public class Match extends FullTextFunction implements Validatable {
     )
     public Match(
         Source source,
-        @Param(name = "field", type = { "keyword", "text" }, description = "Field that the query will target.") Expression field,
+        @Param(
+            name = "field",
+            type = { "keyword", "text", "boolean", "date", "date_nanos", "double", "integer", "ip", "long", "unsigned_long", "version" },
+            description = "Field that the query will target."
+        ) Expression field,
         @Param(
             name = "query",
-            type = { "keyword", "text" },
+            type = { "keyword", "text", "boolean", "date", "date_nanos", "double", "integer", "ip", "long", "unsigned_long", "version" },
```

Comment:
We would allow any combination of parameters - e.g. we allow querying a `date_nanos` field with a `version` value.
I wonder if that's what we want.
The generated documentation will also look interesting - we have 10 possible types for both `field` and `query` parameter - so the generated table for the `match` function signature will have 10*10 = 100 entries?

Should we put some restrictions in place? - e.g. when querying a `version` field you can use `keyword`, `text`, `version` for the `query` param, but not `date_nanos`, `ip`, `boolean` etc

I have not tested with your branch and I am not sure I have seen it tested here, but I am curious what happens if we have:
```
from my-index | where match(ip_field, "aaaa")
```
in `_search` if you try to run a similar `match` query we return a 200 with:
```
{
  "took": 9,
  "timed_out": false,
  "_shards": {
    "total": 13,
    "successful": 12,
    "skipped": 0,
    "failed": 1,
    "failures": [
      {
        "shard": 0,
        "index": "my-index2",
        "node": "v3ts_D6MRYSC0nK5UryhBg",
        "reason": {
          "type": "query_shard_exception",
          "reason": "failed to create query: 'aaaaa' is not an IP string literal.",
          "index_uuid": "-4Bm8TgbS2KaDt-5XktzbA",
          "index": "my-index2",
          "caused_by": {
            "type": "illegal_argument_exception",
            "reason": "'aaaaa' is not an IP string literal."
          }
        }
      }
    ]
  },
  "hits": {
    "total": {
      "value": 0,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  }
}
```

In ES|QL how would this error be surfaced? Should it be a 200 with a returned warning? Or a 400?
We should at least make sure it errors out in a predictable way (not 500).

Reply:
I was too naive implementing parameters options, thanks for raising this.

As I see it:
- Match query can always use a string as a query - it does convert it automatically under the hood. It makes sense that we allow the same thing in ESQL.
- Besides that, the field and the query should match type. We can widen it for numeric types (so an integer can match a field with type double).
- It makes sense to allow just KEYWORD queries for string types

I'm working on this, making this PR draft again until then.

Reply:
I've fixed the allowed parameters so they are compatible. Docs have been updated.

Using casting is also possible to refine the type used:

```
FROM date | WHERE date:"2023-10-23T13:55:01.543"::DATETIME

FROM date_nanos | WHERE nanos:"2023-10-23T13:55:01.543123456Z"::DATE_NANOS

FROM hosts | WHERE ip : "134.45.21.23"::IP

FROM apps | WHERE app_version: "1.2.3"::VERSION
```

Using a string will always do the right thing, independently of the type used:

```
FROM employees | WHERE still_hired:"true"

FROM network_traffic | WHERE bytes_out:"12749081495402663265"

FROM date | WHERE date:"2023-10-23T13:55:01.543"

FROM date_nanos | WHERE nanos:"2023-10-23T13:55:01.543123456Z"

FROM hosts | WHERE ip : "134.45.21.23"

FROM apps | WHERE app_version: "1.2.3"
```

so casting is probably something we can get rid of - but adding here for completeness and discussion 👍